File: update_images.py
import re
import time
from duckduckgo_search import DDGS

import sys
sys.stdout.reconfigure(encoding='utf-8')

FILE_PATH = r'c:\Users\Asus\Downloads\BRAVEECOMUISTITCH\stitch_brave_ecom_pvt_ltd\sunray_system\app\api\mall\products\route.ts'

with open(FILE_PATH, 'r', encoding='utf-8') as f:
    content = f.read()

# Pattern aiming for: ... id: "MALL-PRD...", name: "Apple AirPods...", ... image: getImg('earbuds', 1)
pattern = re.compile(r'("?id"?:\s*"(MALL-PRD-\d+)"[\s\S]*?"?name"?:\s*"([^"]+)"[\s\S]*?"?image"?:\s*)"([^"]+)"')

matches = pattern.finditer(content)
matches_list = list(matches)

if not matches_list:
    print("No matches found!")
    sys.exit(1)

print(f"Found {len(matches_list)} products. Starting image fetch...")

new_content = content
ddgs = DDGS()

for i, match in enumerate(matches_list):
    prefix = match.group(1)
    prod_id = match.group(2)
    name = match.group(3)
    old_img_call = match.group(4)
    
    query = f"{name} high quality product photo isolated white background"
    print(f"[{i+1}/{len(matches_list)}] Searching for: {query}")
    
    try:
        results = ddgs.images(query, max_results=1)
        if results and len(results) > 0:
            image_url = results[0]['image']
            image_url = image_url.replace('"', "%22")
            
            # Since content is changing, replace precisely the exact string for this specific product ID match
            # To be safe, we'll replace the full matched substring
            full_match_str = match.group(0)
            replacement_str = prefix + f'"{image_url}"'
            new_content = new_content.replace(full_match_str, replacement_str)
            print(" -> Found: " + image_url)
        else:
            print(" -> No results found!")
    except Exception as e:
        print(f" -> Error: {e}")
    
    time.sleep(1.5) # avoid rate limits

with open(FILE_PATH, 'w', encoding='utf-8') as f:
    f.write(new_content)

print("Finished updating images!")

File: update_images.js
const fs = require('fs');
const google = require('googlethis');
const path = require('path');

const FILE_PATH = path.join(__dirname, 'app', 'api', 'mall', 'products', 'route.ts');

async function main() {
    let content = fs.readFileSync(FILE_PATH, 'utf-8');

    // Pattern to grab lines with `image: getImg(...)`
    // We want to capture the prefix, the product name, and the getImg call exactly.
    // Example: { id: "MALL-PRD-00007", name: "Realme 13 Pro+ 5G", ... image: getImg('smartphone', 7)
    const regex = /(id:\s*"MALL-PRD-\d+".*?name:\s*"([^"]+)".*?image:\s*)(getImg\([^)]+\))/g;

    let match;
    const matches = [];
    while ((match = regex.exec(content)) !== null) {
        matches.push({
            full: match[0],
            prefix: match[1],
            name: match[2],
            oldImg: match[3]
        });
    }

    if (matches.length === 0) {
        console.log("No more getImg placeholders found!");
        return;
    }

    console.log(`Found ${matches.length} products to update. Starting Google Images search...`);

    for (let i = 0; i < matches.length; i++) {
        const item = matches[i];
        const query = `${item.name} high res product ecommerce transparent background white bg -site:pinterest.com`;
        console.log(`[${i + 1}/${matches.length}] Searching: ${item.name}`);

        try {
            const images = await google.image(query, { safe: false });
            if (images && images.length > 0) {
                // Pick the first reliable image (often amazon, flipkart, or official sites)
                let bestImg = images[0].url;

                // If it's a huge base64 or completely broken URL, try next
                if (bestImg.length > 500) {
                    bestImg = images[1]?.url || bestImg;
                }

                console.log(` -> Found: ${bestImg}`);

                const replacement = `${item.prefix}"${bestImg}"`;
                content = content.replace(item.full, replacement);

                // Write incrementally so we save progress in case it crashes
                fs.writeFileSync(FILE_PATH, content, 'utf-8');
            } else {
                console.log(` -> No results for ${item.name}`);
            }
        } catch (e) {
            console.error(` -> Error fetching ${item.name}: ${e.message}`);
        }

        // Anti-ratelimit delay
        await new Promise(r => setTimeout(r, 1500));
    }

    console.log("Image update complete!");
}

main().catch(console.error);
